### ISSUE TITLE
Allow uploading of zip containing notebook with resources

### ISSUE THREAD

**Issue by nitind** [jupyter-incubator/dashboards_server#79]
The ability to upload a notebook to the dashboards server does not allow for resources to go with it, such as widget files and data. Allow a zip file to be uploaded and its resources unpacked and served without clobbering the resources of other notebooks.

---

**Structure:**  *[updated 2016-02-26]*

```
<notebook_name>/
    index.ipynb
    urth_components/
        ...
    stylesheet.css
    image.jpg
```

---

**Client/Server Tasks:**
* [X] Support plain and bundled dashboards.
    * Plain & bundled should look same in UI.
    * Don't show `ipynb` extension.
* [X] Support GET of bundled resources (everything but the notebook itself).
* [x] Look for `index.ipynb` and always render for that dashboard for the dir, whether inside or outside of a "bundled" dashboard.
    * Look for `index` instead of `__notebook__`
* [x] Load declwidgets from within bundled dashboard
* [x] Create example installable through Makefile. Should test declwidgets, loading of local resources (CSS, images, etc).
* [x] Support POST of bundled dashboards into `data` dir
* [x] Create tests
    * Test loading `index.ipynb` (with/without extension). We should **not** get full contents back.
    * etc
* [x] Update README to describe use of `index.ipynb` in a dir and local resources.

**Comment by parente** [comment]
This can be done in conjunction with support dynamic imports from declarative widgets.

**Comment by parente** [comment]
I setup some time to chat on Monday about getting the design right for this:

* Should we support both plain ipynb uploads and zip uploads?
* When we support zip uploads, how do the zips get expanded? Into their own directories under `data/`?
* When the zip upload contains web assets that need to be made accessible to the dashboard frontend at runtime, how are they mapped to a URL path? Are all assets under `data/*/**` subdirectories web accessible? Do we have to check file extensions to see if the asset is another notebook that should be rendered or anything else that should be served up with the right mimetype? (Yuck.) Or do we invent a rule like "everything under the relative path `static/` in the zip will be made web accessible? (Then the bundler can know where to put stuff.) 
* When the bundle contains polymer widgets, I believe it needs to be self-contained with its own copy of polymer, urth_widgets, etc. like we did with the local dashboard deploy to avoid conflicts of versions across dashboards and to satisfy the requirement that all polymer components come from a single path. What parts of declarative widgets support do we need to keep in the core here (e.g., the JS shims) and what parts need to go and become part of the bundle upload (e.g., the install of polymer components that is currently happening in the Makefile for demo purposes?)

@jhpedemonte @dalogsdon

**Comment by parente** [comment]
We had a chat about the above in appear.

> Should we support both plain ipynb uploads and zip uploads?

Yes. The bundler might end up sending zips exclusively from now on, but we want to keep the API open for other clients to send notebooks to the server in various ways.

> When we support zip uploads, how do the zips get expanded? Into their own directories under data/?

From the `dashboards_bundler`, *Deploy as &rarr; Dashboard on Dashboard Server*, `foo.zip` would have the following internal structure if using the code we already have for *Deploy as &rarr; Local dashboard* today.

```
foo.ipynb
static/
	urth_components/  # components installed using declarative_widgets
        urth_widgets/ # copy of declarative widgets
        # other supporting assets
some_widget.css # other frontend files associated with the notebook
some_widget.html
```

As it stands, any custom widgets or whatnot associated with the notebook using the pre markup (like in the meetup outreach demo) are not put in static. This way, relative references work in the notebook server or when deployed.

On the dashboard server, the above would be `POST`ed to `/foo` and expanded into the equivalent subfolder, with `foo.ipynb` rewritten as `index.ipynb` (if it's not already called `index.ipynb` in the zip). Overwrite becomes a bit more complicated now since a directory will exist. (It should be wiped and recreated since we assume zip files are self-contained. See below.)

> When the zip upload contains web assets that need to be made accessible to the dashboard frontend at runtime, how are they mapped to a URL path?

Need to figure out:

* If just `static/` gets mapped to a web url, what about the other notebook-associated assets that were in relative paths?
* What path does `dashboard.js#initWidget` pass to declarative widgets to point to the proper `static/` directory for the notebook / dashboard?

> When the bundle contains polymer widgets, I believe it needs to be self-contained with its own copy of polymer, urth_widgets, etc.

We all agreed. We've been down the road of trying to ship some assets separately and always run into trouble that way.

**Comment by jhpedemonte** [comment]
Thinking of calling the notebook something unique to us, instead of `index.ipynb`. I'm thinking of the case where a user wants to have a subdir with an `index.ipynb` in order to organize other notebooks in that directory. We don't want to make that appear as a single notebook.

Maybe call it `__notebook__.ipynb` or something along those lines.

**Comment by jhpedemonte** [comment]
I updated description above with the structure that we are currently working with.

**Comment by jhpedemonte** [comment]
Created client/server tasks in description.

**Comment by parente** [comment]
We talked about making index.ipynb consistently render in place of a file listing in subdirs whether they're bundles or not.

**Comment by jhpedemonte** [comment]
> foo.zip would have the following internal structure

Actually, the way we're currently handling this is that there is no `static` directory within the bundled dashboard. It should follow the structure posted in the description above.

> If just static/ gets mapped to a web url, what about the other notebook-associated assets that were in relative paths?

We expose all files in the bundled dashboard dir. Notebooks themselves will render (so user cannot get full notebook contents), but for other files, we return content.

**Comment by jhpedemonte** [comment]
Latest commit adds support for uploading zip.

**Comment by dalogsdon** [comment]
The changes to when we render the index notebook will revert #95.

**Comment by parente** [comment]
Yep. Live and learn. (And something tells me we'll be back here again if/when other people start looking at using it...)

**Comment by parente** [comment]
Closed by PR #122. Any issues related to bundling, doc, etc. will be done as follow-ons.

![deploy2](https://cloud.githubusercontent.com/assets/153745/13505317/aff393a0-e146-11e5-8be9-5624ff203552.gif)
